### PR TITLE
[expo-go] Account for snack when setting query params

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
When we receive a URL from a deep link, we strip the query parameters and only add back the ones relevant to expo updates. This causes the parameter necessary for snack to be lost.

# How
Check if the URL contains `snack-channel` and append it at the end. I've also made `getMainModuleName` nullable and provided a default of `main`. This makes the behaviour consistent with ios

# Test Plan
Expo go, the URL is parsed correctly

### CODE REVIEW

File: apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/Kernel.kt [hunk 1]
```diff
@@ -524,7 +526,7 @@ class Kernel : KernelInterface() {
       )
     }
 
-    // transfer the expo-updates query params: runtime-version, channel-name
+    // transfer the expo-updates query params: runtime-version, channel-name )
```
Comment (Kudo) [inline apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/Kernel.kt:529]: unused tailed `)`?

File: apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/Kernel.kt [hunk 2]
```diff
@@ -614,15 +620,12 @@ class Kernel : KernelInterface() {
         manifestUrl,
         object : AppLoaderCallback {
           override fun onOptimisticManifest(optimisticManifest: Manifest) {
-            kernelScope.launch {
-              sendOptimisticManifestToExperienceActivity(
-                optimisticManifest
-              )
-            }
+            Exponent.instance
+              .runOnUiThread { sendOptimisticManifestToExperienceActivity(optimisticManifest) }
```
Comment (Kudo) [inline apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/Kernel.kt:624]: is that intentional to rollback `Exponent.instance.runOnUiThread`? i was thinking that you try to remove this singleton by kernelScope?

Comment (alanjhughes) [reply apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/Kernel.kt:624]: I was just debugging something and forgot to revert this 😅

**Review by Kudo — APPROVED** [review]

File: packages/expo-manifests/android/src/main/java/expo/modules/manifests/core/Manifest.kt [hunk 3]
```diff
@@ -91,7 +91,7 @@ abstract class Manifest(protected val json: JSONObject) {
   abstract fun getSlug(): String?
 
   fun getDebuggerHost(): String = getExpoGoConfigRootObject()!!.require("debuggerHost")
-  fun getMainModuleName(): String = getExpoGoConfigRootObject()!!.require("mainModuleName")
+  fun getMainModuleName(): String = getExpoGoConfigRootObject()?.require("mainModuleName") ?: "main"
```
Comment (wschurman) [inline packages/expo-manifests/android/src/main/java/expo/modules/manifests/core/Manifest.kt:94]: I tried doing some searching for this on iOS and couldn't find anything. Can you elaborate in the summary about the iOS behavior this is equivalent to? (might be worth separating out this change from the snack URL changes unless they're interrelated. I'm happy to try to help dig into the `mainModuleName` ones like we did a week ago)

**Review by wschurman — APPROVED** [review]
Eh, nvm elaborating on iOS, that code is super old, no need to dig. This looks good.

File: packages/expo-manifests/android/src/main/java/expo/modules/manifests/core/Manifest.kt (hunk 3 above)
Comment (wschurman) [reply packages/expo-manifests/android/src/main/java/expo/modules/manifests/core/Manifest.kt:94]: Optional: Could fix this in the same way as we did the previous fix and only call this method in development mode. It looks like https://github.com/expo/expo/pull/31395 added additional calls in non-development mode.

```diff
--- a/apps/expo-go/android/expoview/src/main/java/host/exp/exponent/experience/ReactNativeActivity.kt
+++ b/apps/expo-go/android/expoview/src/main/java/host/exp/exponent/experience/ReactNativeActivity.kt
@@ -378,7 +378,11 @@ abstract class ReactNativeActivity :
     val nativeHost = ExpoGoReactNativeHost(
       application,
       instanceManagerBuilderProperties,
-      manifest!!.getMainModuleName()
+      if (delegate.isDebugModeEnabled) {
+        manifest!!.getMainModuleName()
+      } else {
+        null
+      }
     )

     val devBundleDownloadListener = ExponentDevBundleDownloadListener(progressListener)
```

I agree overall though that the behavior should match on both platform. I can't seem to find the equivalent on iOS though. 

`mainModuleName` is only available during development: https://github.com/expo/expo/blob/main/packages/%40expo/cli/src/start/server/middleware/ManifestMiddleware.ts#L286, so it seems that android may have been stricter about this constraint on the client than iOS? Either way, my approval still stands with the code as-is (making this nullable) assuming it matches iOS; Expo Go is a fairly archaic codebase and trying to do code archaeology in it usually isn't worth it.

**Review by wschurman — APPROVED** [review]

File: packages/expo-manifests/android/src/main/java/expo/modules/manifests/core/Manifest.kt (hunk 3 above)
Comment (alanjhughes) [reply packages/expo-manifests/android/src/main/java/expo/modules/manifests/core/Manifest.kt:94]: Here is where we provide `main` as the default on iOS https://github.com/expo/expo/blob/b2bca062a545f51961178f6ef0cf75ce20ba7d66/apps/expo-go/ios/Exponent/Kernel/ReactAppManager/EXReactAppManager.mm#L551-L570

**Pull request merged**